Use intrusive_ptr for LazyTensor

wconstab opened the pull request:
Summary:
Refactors the whole codebase to use LazyTensorPtr (defined as c10::intrusive_ptr) to enable XLA to use a derived class XlaLazyTensor and override functionality.

this PR is just the first step, and we will need to add a factory class that XLA can override in their backend to actually hook up their derived tensor class.

Parallel PR on lazy_tensor_staging: #73429

Test Plan: tested via lazy_tensor_staging test_ptltc and torchbench and CI

Differential Revision: D34481918

ezyang commented on torch/csrc/lazy/backend/backend_device.cpp:
```diff
@@ -56,7 +56,9 @@ c10::Device backendDeviceToAtenDevice(const BackendDevice& device) {
 c10::optional<BackendDevice> GetBackendDevice(const at::Tensor& tensor) {
   if (auto lt = TryGetLtcTensor(tensor)) {
-    return lt.GetDevice();
+    if (*lt) {
```
this looks a little questionable!

wconstab replied:
hah yea, you caught me. it is legit,  i think, but indicates a necessary refactor/cleanup.

old world:
LazyTensor::operator bool() -- returns false if LazyTensor.data is null

new world:
LazyTensorPtr::operator bool() -- returns false if LazyTensorPtr is uninitialized
LazyTensor::operator bool() -- returns false if LazyTensor.data is null

we should delete the 'Data' container class inside LazyTensor and rely entirely on unitialized LazyTensorPtr as the only way to represent a 'null' LazyTensor.

But to decouple that big change, I wanted to just assume Data() is never allowed to be null anymore but otherwise leave the API as-is.  I think I can make this change now (assert LazyTensor.data() is never null) and then I could remove this check.

wconstab committed:
Use intrusive_ptr for LazyTensor (#73445)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/73445

Refactors the whole codebase to use LazyTensorPtr (defined as c10::intrusive_ptr) to enable XLA to use a derived class XlaLazyTensor and override functionality.

this PR is just the first step, and we will need to add a factory class that XLA can override in their backend to actually hook up their derived tensor class.

Parallel PR on lazy_tensor_staging: #73429

Test Plan: tested via lazy_tensor_staging test_ptltc and torchbench and CI

Differential Revision: D34481918

fbshipit-source-id: 0bd874a530e3dd73affe206911d7b907239bf839

ezyang commented on torch/csrc/lazy/core/tensor_impl.h:
```diff
@@ -50,7 +51,7 @@ class TORCH_API LTCTensorImpl final : public c10::TensorImpl {
   void setup_size_properties();
 
-  LazyTensor tensor_;
+  LazyTensorPtr tensor_;
```
It took me a few minutes of thinking to figure out why this is a pointer, because traditionally fields in a TensorImpl are stored inline, and turning this into an intrusive pointer introduces an indirection for seemingly no benefit.

What the benefit seems to be, is that LazyTensor may be specialized to XLATensor (in the case of XLA's backend), and now you need the indirection to be able to swap out backends.

That is all very well and fine but... does it really need to be a refcounted pointer? A `unique_ptr` to a class with a virtual destructor seems like it would be sufficient as well, since multiple references to the old LazyTensor shouldn't have been possible either.

ezyang replied:
(psst, a comment here about why the indirection helps would be a good idea!)

ezyang commented on torch/csrc/lazy/core/tensor_impl.cpp:
```diff
@@ -73,14 +77,14 @@ LTCTensorImpl::LTCTensorImpl(LazyTensor&& tensor)
-void LTCTensorImpl::set_tensor(const LazyTensor& lazy_tensor) {
-  tensor_ = lazy_tensor;
+void LTCTensorImpl::set_tensor(const LazyTensorPtr& lazy_tensor) {
+  tensor_ = c10::make_intrusive<LazyTensor>(*lazy_tensor);
```
Case in point: this is super puzzling under the current typing of LazyTensorPtr (you're doing a copy of LazyTensor when you have a perfectly good reference counted pointer). But if it was a UniquePtr (and you took a pointer to LazyTensor directly) the copy would make more sense.

ezyang commented on torch/csrc/lazy/core/tensor_impl.cpp:
```diff
@@ -65,6 +65,10 @@ C10_REGISTER_GUARD_IMPL(Lazy, LTCGuardImpl);
 
 }  // namespace
 
+// TODO(whc) when do we want to clone vs share?
```
Never share? :)

ezyang reviewed (approved):
In the spirit of not blocking refactors, here's your stamp, but there is something fishy going on here...

wconstab commented:
@ezyang thanks, I may go ahead and merge this as is but I want to take it further in some additional PRs.

I think there are 2 directions I could go-

a more modest cleanup
 * we kill the 'Data' class and its shared_ptr inside LazyTensor and fold its stuff into LazyTensor directly
 * a bit more polish on the LazyTensor ptr in LTCTensorImpl: maybe make it a unique_ptr after all, maybe change the naming conventions
 
or more radical
  * do we actually need a pointer inside LTCTensorImpl?  We do want to allow indirection somewhere, but could we let XLA provide their own XLATensorImpl instead?
  * I could use a brainstorm session on how TensorImpls work and what you're allowed to do with them
